packages/horizon/contracts/payments/collectors/TAPCollector.sol

```diff
@@ -137,4 +173,74 @@ contract TAPCollector is EIP712, GraphDirectory, ITAPCollector {
                 )
             );
     }
+
+    /**
+     * @notice Verify the proof provided by the payer authorizing the signer
+     * @param proof The proof provided by the payer authorizing the signer
+     * @param proofDeadline The deadline by which the proof must be verified
+     * @param signer The signer to be authorized
+     */
+    function verifyAuthorizedSignerProof(bytes calldata proof, uint256 proofDeadline, address signer) private view {
+        // Verify that the proofDeadline has not passed
+        require(
+            proofDeadline > block.timestamp,
+            TAPCollectorInvalidSignerProofDeadline(proofDeadline, block.timestamp)
+        );
+
+        // Generate the hash of the payer's address
+        bytes32 messageHash = keccak256(abi.encodePacked(block.chainid, proofDeadline, msg.sender));
+
+        // Generate the digest to be signed by the signer
+        bytes32 digest = MessageHashUtils.toEthSignedMessageHash(messageHash);
+
+        // Verify that the recovered signer matches the expected signer
+        require(ECDSA.recover(digest, proof) == signer, TAPCollectorInvalidSignerProof());
+    }
+
+    /**
+     * @notice See {ITAPCollector.collect}
+     */
+    function _collect(
+        IGraphPayments.PaymentTypes paymentType,
+        address payer,
+        SignedRAV memory signedRAV,
+        uint256 dataServiceCut
+    ) private returns (uint256) {
+        address dataService = signedRAV.rav.dataService;
+        address receiver = signedRAV.rav.serviceProvider;
+
+        uint256 tokensRAV = signedRAV.rav.valueAggregate;
+        uint256 tokensAlreadyCollected = tokensCollected[dataService][receiver][payer];
+        require(
+            tokensRAV > tokensAlreadyCollected,
+            TAPCollectorInconsistentRAVTokens(tokensRAV, tokensAlreadyCollected)
+        );
+
+        uint256 tokensToCollect = tokensRAV - tokensAlreadyCollected;
+        uint256 tokensDataService = tokensToCollect.mulPPM(dataServiceCut);
+
+        if (tokensToCollect > 0) {
+            _graphPaymentsEscrow().collect(
+                paymentType,
+                payer,
+                receiver,
+                tokensToCollect,
+                dataService,
+                tokensDataService
+            );
+            tokensCollected[dataService][receiver][payer] = tokensRAV;
```

Comment:
The call to `_graphPaymentsEscrow().collect` is to a trusted contract so I don't see a direct risk of reentrancy, but it'd  still be cleaner/safer to move this before that call